concurrent_map.go

```diff
@@ -290,7 +316,27 @@ func (m ConcurrentMap) Keys() []string {
 	return keys
 }
 
-//Reviles ConcurrentMap "private" variables to json marshal.
+// FindOldest search for the oldest key in the key slice.
+func (m ConcurrentMap) FindOldest() string {
+	keys := []string{}
+	for i := 0; i < ShardCount; i++ {
+		shard := m[i]
+		if len(shard.mapKeys) > 0 {
+			shard.Lock()
+			keys = append(keys, shard.mapKeys[0])
+			shard.Unlock()
+		}
+	}
+	return keys[0]
```

Comment:
I see you iterate over all shards and gather the oldest in each and add those to keys, but you always return the first one in keys. 
This does not cover the case when shard7 reached it's capacity, but the "oldest one" you remove is from shard0, so you still do not have capacity in shard7.

So I think if you just remove the oldest from the shard you plan to add to, if there is no empty space in the shard then it should be fine

Reply:
Yes, but the cacher interface is exposing a `RemoveOldest` method which does not take into consideration the sharding mechanism. To remove the last item from the shard first we have to provide a key in order to find the shard the key belongs to. But since the `RemoveOldest` method signature does not accept a parameter, it's not possible to implement this method in this manner. This is why i'm always returning the last element without considering the shard from which i want to return the last element.

I can remove the oldest item from the shard I wish to insert a new value, but the question is how we should adapt the `RemoveOldest` method to the current sharding scheme?

Reply:
remove oldest can just be empty function, if you take care of removing the oldest when needed (at addition time when the shard map is full)